backend/src/org/akvo/dash/transformation.clj

```diff
@@ -1,225 +1,157 @@
-;;   {:columns [{:title "Name" :type "STRING" :values ["1" "2" "3"]}
-;;              {:title "Age" :type "STRING" :values ["1" "2" "3"]}]}
-;;   "Age"
-;;   "NUMBER"))
-
-
-
-
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;; Playing with traversy lenses
+  (:require [clojure.java.jdbc :as jdbc]
+            [clojure.string :as str]
+            [hugsql.core :as hugsql]
+            [org.akvo.dash.transformation.engine :as engine]
+            [org.akvo.dash.util :refer (squuid gen-table-name)]))
+
+;; TODO: Potential change op-spec validation `core.spec`
+
+(def ops-set (set (keys engine/available-ops)))
+(def on-error-set #{"fail" "default-value" "delete-row"})
+(def type-set #{"number" "text" "date"})
+(def sort-direction-set #{"ASC" "DESC"})
+
+(hugsql/def-db-fns "org/akvo/dash/transformation.sql")
+
+
+(defn- throw-invalid-op
+  [op-spec]
+  (throw (Exception. (str "Invalid operation " op-spec))))
+
+(defn required-keys
+  [{:strs [op args onError] :as op-spec}]
+  (or (boolean (and (ops-set op)
+                    (not-empty args)
+                    (re-find #"c\d+" (args "columnName"))
+                    (on-error-set onError)))
+      (throw-invalid-op op-spec)))
+
+(defmulti valid-op?
+  (fn [op-spec]
+    (keyword (op-spec "op"))))
+
+(defmethod valid-op? :default
+  [op-spec]
+  (throw-invalid-op op-spec))
+
+(defmethod valid-op? :core/change-datatype
+  [{:strs [op args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                 (type-set (args "newType"))
+                 (not-empty (args "defaultValue"))
+                 (if (= "date" (args "newType"))
+                   (not-empty (args "parseFormat"))
+                   true)))
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/change-column-title
+  [{:strs [args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                    (not-empty (args "columnTitle"))))
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/sort-column
+  [{:strs [args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                    (sort-direction-set (args "sortDirection"))))
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/remove-sort
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/filter
+  [op-spec]
+  true)
+
+(defmethod valid-op? :core/to-titlecase
+  [op-spec]
+  (required-keys op-spec))
+
+(defmethod valid-op? :core/to-lowercase
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/to-uppercase
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/trim
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/trim-doublespace
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defn valid?
```

Comment:
Agree. those started as predicate functions returning true/false and then the implementation went in a different direction :wink: